libnvme-rs/src/nvme_uri.rs

```diff
@@ -0,0 +1,277 @@
+            nr_write_queues: 0,
+            nr_poll_queues: 0,
+            tos: -1,
+            duplicate_connect: false,
+            disable_sqflow: false,
+            hdr_digest: false,
+            data_digest: false,
+        };
+        let ret = unsafe {
+            crate::nvmf_add_ctrl(h, nvme_ctrl, &cfg, cfg.disable_sqflow)
+        };
+        if ret != 0 {
+            return Err(NvmeError::AddCtrlrError {
+                rc: ret,
+            });
+        }
+
+        let mut retries = 10;
+        let mut all_nvme_devices = Vec::<String>::new();
+        loop {
+            std::thread::sleep(Duration::from_millis(1000));
+
+            let r =
+                NvmeRoot::new(unsafe { crate::nvme_scan(std::ptr::null()) });
+            let hostiter = NvmeHostIterator::new(&r);
+            for host in hostiter {
+                let subsysiter = NvmeSubsystemIterator::new(host);
+                for subsys in subsysiter {
+                    let cstr = unsafe {
+                        std::ffi::CStr::from_ptr(crate::nvme_subsystem_get_nqn(
+                            subsys,
+                        ))
+                    };
+                    if cstr.to_str().unwrap() != self.subsysnqn {
+                        continue;
+                    }
+                    let nsiter = NvmeNamespaceIterator::new(subsys);
+                    for ns in nsiter {
+                        all_nvme_devices.push(format!(
+                            "/dev/{}",
+                            unsafe {
+                                std::ffi::CStr::from_ptr(
+                                    crate::nvme_ns_get_name(ns),
+                                )
+                            }
+                            .to_str()
+                            .unwrap()
+                        ));
+                    }
+                }
+            }
+
+            retries -= 1;
+            if retries == 0 || !all_nvme_devices.is_empty() {
+                break;
+            }
+        }
+
+        Ok(all_nvme_devices)
+    }
+
+    /// Disconnect from NVMe target
+    /// Returns number of controllers disconnected
+    pub fn disconnect(&self) -> Result<usize, NvmeError> {
+        let r = NvmeRoot::new(unsafe { crate::nvme_scan(std::ptr::null()) });
+        let mut i = 0;
+        let hostiter = NvmeHostIterator::new(&r);
+        for host in hostiter {
+            let subsysiter = NvmeSubsystemIterator::new(host);
+            for subsys in subsysiter {
+                let cstr = unsafe {
+                    std::ffi::CStr::from_ptr(crate::nvme_subsystem_get_nqn(
+                        subsys,
+                    ))
+                };
+                if cstr.to_str().unwrap() != self.subsysnqn {
+                    continue;
+                }
+                let ctrlriter = NvmeCtrlrIterator::new(subsys);
+                for c in ctrlriter {
+                    if unsafe { crate::nvme_disconnect_ctrl(c) == 0 } {
```

Comment:
What does != 0 mean here? cant be found? or, can be found but can't disconnect? or.... ? if the disconnect call would yield EAGAIN or EBUSY, then yeah I can imagine we MAY want to retry but perhaps it's better to handle that further up, and let the consumer decide -- not the library.

Reply:
This basically writes `1` to `/sys/class/nvme/{}/delete_controller` so `!=0` would an IO error of some description. I'll try to pass the error back up.

Reply:
I've passed the error back up as just the raw return code as there's no easy way to tell where it failed.